kernel/examples/inspect-table/src/main.rs

```diff
@@ -43,39 +46,161 @@ enum Commands {
     },
 }
 
-fn main() {
-    let cli = Cli::parse();
-    let path = std::fs::canonicalize(PathBuf::from(cli.path));
-    let Ok(path) = path else {
-        println!("Couldn't open table: {}", path.err().unwrap());
-        return;
-    };
-    let Ok(url) = url::Url::from_directory_path(path) else {
-        println!("Invalid url");
-        return;
+fn main() -> ExitCode {
+    env_logger::init();
+    match try_main() {
+        Ok(()) => ExitCode::SUCCESS,
+        Err(e) => {
+            println!("{e:#?}");
+            ExitCode::FAILURE
+        }
+    }
+}
+
+enum Action {
+    Metadata(delta_kernel::actions::Metadata, usize),
+    Protocol(delta_kernel::actions::Protocol, usize),
+    Remove(delta_kernel::actions::Remove, usize),
+    Add(delta_kernel::actions::Add, usize),
+}
+
+impl Action {
+    fn row(&self) -> usize {
+        match self {
+            Action::Metadata(_, row) => *row,
+            Action::Protocol(_, row) => *row,
+            Action::Remove(_, row) => *row,
+            Action::Add(_, row) => *row,
+        }
+    }
+}
+
+fn fields_in(field: &StructField) -> usize {
+    if let DataType::Struct(ref inner) = field.data_type {
+        let mut count = 0;
+        for field in inner.fields() {
+            count += fields_in(field);
+        }
+        count
+    } else {
+        1
+    }
+}
+
+struct LogVisitor {
+    actions: Vec<Action>,
+    add_offset: usize,
+    remove_offset: usize,
+    protocol_offset: usize,
+    metadata_offset: usize,
+    previous_rows_seen: usize,
+}
+
+impl LogVisitor {
+    fn new(log_schema: &SchemaRef) -> LogVisitor {
+        let mut offset = 0;
+        let mut add_offset = 0;
+        let mut remove_offset = 0;
+        let mut protocol_offset = 0;
+        let mut metadata_offset = 0;
+        for field in log_schema.fields() {
+            match field.name().as_str() {
+                "add" => add_offset = offset,
+                "remove" => remove_offset = offset,
+                "protocol" => protocol_offset = offset,
+                "metaData" => metadata_offset = offset,
+                _ => {}
+            }
+            offset += fields_in(field);
+        }
+        LogVisitor {
+            actions: vec![],
+            add_offset,
+            remove_offset,
+            protocol_offset,
+            metadata_offset,
+            previous_rows_seen: 0,
+        }
+    }
+}
+
+impl DataVisitor for LogVisitor {
+    fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()> {
+        for i in 0..row_count {
+            if let Some(path) = getters[self.add_offset].get_opt(i, "add.path")? {
```

Comment:
I noticed the `DataVisitor::visit` has a TODO for a better doc comment.  This may be a good opportunity to add that. In particular, I want to know what getters are supposed to look like. Even a pointer to docs for `EngineData::extract` would be handy. I'm also interested in how `TypedGetData::get`/`TypedGetData::get_opt` work wrt the field name. Do we always have to start at the top level field (ie `add.path`), or could we have done `path` here since getters only stores leaves.

Reply:
Yep, added a better comment there.

Note that this is an _internal_ trait that connector/engine authors don't need to worry about.

Re `add.path` vs. `path`, that's actually _only_ for debugging purposes. The `getters` already are only for leaf nodes in the schema passed to `extract`. So we'll just return an `Err` with `add.path` in the message if we didn't find a `String` in the getter. It helps _a lot_ when you're working on any code that touches this stuff.

Reply:
Aha, so just a label for debuggability, makes sense